### ISSUE TITLE
Option to disable top-scope variable warning for $facts and $trusted hashes

### ISSUE THREAD

**Issue by elyscape** [rodjek/puppet-lint#382]
Since Puppet 3.5, `$facts` and `$trusted` have been available if [`trusted_node_data`](https://docs.puppetlabs.com/references/3.5.latest/configuration.html#trustednodedata) is enabled and cannot be overridden in any scope unless for some reason [`immutable_node_data`](https://docs.puppetlabs.com/references/3.5.latest/configuration.html#immutablenodedata) is disabled as well.  Despite this, puppet-lint flags uses of `$facts` and `$trusted` as [potential scoping issues](http://puppet-lint.com/checks/variable_scope/).  Turning off that check is undesirable because it's still useful for finding other possible scoping problems, but putting `$::facts` or `$::trusted` is unnecessary for codebases that only run on systems with `trusted_node_data` and `immutable_node_data` enabled.  Furthermore, this is the default configuration for PE3.7 and will be the default for Puppet 4.  A possible solution to this would be a flag to disable this warning but only for these variables.  To illustrate:

Current behavior:
```shellsession
$ cat -b test.pp
     1  class test {
     2    $good = $facts['puppetversion']
     3    $bad = $puppetversion
     4    $standard = $::puppetversion
     5  }
$ puppet-lint test.pp
ERROR: test not in autoload module layout on line 1
WARNING: top-scope variable being used without an explicit namespace on line 2
WARNING: top-scope variable being used without an explicit namespace on line 3
WARNING: class not documented on line 1
```
Desired behavior:
```shellsession
$ cat -b test.pp
     1  class test {
     2    $good = $facts['puppetversion']
     3    $bad = $puppetversion
     4    $standard = $::puppetversion
     5  }
$ puppet-lint --no-immutable_variables-warning test.pp
ERROR: test not in autoload module layout on line 1
WARNING: top-scope variable being used without an explicit namespace on line 3
WARNING: class not documented on line 1
```

**Comment by sathieu** [comment]
I don't agree. Those are top-scope facts and should be written accordingly.

**Comment by elyscape** [comment]
When `trusted_node_data` is enabled, `immutable_node_data` is by default enabled. When `immutable_node_data` is enabled, assigning to `$facts` and `$trusted` is forbidden in any scope. Both `trusted_node_data` and `immutable_node_data` will default to enabled in Puppet 4 and already default to enabled in PE3.7.  As such, it makes sense to offer the option to turn off scope warnings for those specific variables.

**Comment by sathieu** [comment]
Yes I know that. But immutability and scope are two different concepts. Using $::hostname doesn't prevent mutation, even if it's top scope.

My understanding of the style guide is to be explicit about where the variable comes from, not about it's immutability.

Anyway, I'm not the reviewer here. Waiting for the project owner.

**Comment by elyscape** [comment]
That's my point, though. It makes it immutable in the top scope and causes a compilation error if you try to assign to those names in any other scope.

**Comment by zleswomp** [comment]
https://docs.puppet.com/puppet/latest/reference/lang_reserved.html#reserved-variable-names

`$facts` and `$trusted` are reserved and will not be modified in local scope.  More code will begin using these variables without the topscope reference, and to leave this lint check in place as is, feels a bit like a tail wagging the dog.  I would also like to see this project allow for exceptions to the topscope rule so that I may follow the puppet style as is evolves, rather than code style being anchored by the puppet-lint project.

I suspect that if this change is not made, more people will be just disabling the check instead of modifying their code, and that feels a bit counter to the goals here.

**Comment by tehmaspc** [comment]
Good discussion here; I would love some resolution to this as well.

I would agree w/ @zleswomp and @elyscape due to the changes in Puppet 4. I'd rather not have to disable the check though I could. Mainly I want to set good standards for the rest of my team :)

Thanks!

**Comment by mmckinst** [comment]
This was actually found in November 2015 and fixed in January 2015 (PR #362), however at the time this issue was filed, the latest version on rubygems did not have the fix. Unless people had their Gemfile configured to get puppet-lint from git, its easy to see why people were running in to the issue.

puppet-lint 2.0.0 was released on rubygems June 22, 2016 and has the fix in it.

If you're running in to this problem today (like I was), its because you're inadvertently running puppet-lint 1.1.0 instead of the latest 2.0.0. There's a few reasons that might be happening:

1. You've hardcoded the version of puppet-lint in Gemfile like one of the below.

```
# bad
gem 'puppet-lint', '~>1.0'

# good
gem 'puppet-lint', '~>2.0'
```

2. You're using another gem that depends on puppet-lint ~>1.0. 

We had a few 3rd party puppet-lint gems like puppet-lint-unquoted_string-check, puppet-lint-version_comparison-check, puppet-lint-variable_contains_upcase, etc we were using in addition to the main puppet-lint gem. We discovered one of them said it was only compatible with puppet-lint ~>1.0, while all the rest were compatible with puppet-lint < 3.0, >= 1.0. bundler was choosing the highest version supported by all of them which ended up being 1.1.0. To check for this you can run the following command (see the below output for an example of what it looks like, notice the line for puppet-lint-variable_contains_upcase): `bundle exec gem dependency --reverse-dependencies puppet-lint | fgrep '(puppet-lint '`. 

To fix this you'll either need to fix the incompatabile gem or stop using it and remove it from your Gemfile. In our case it was puppet-lint-variable_contains_upcase that was causing the issue and @rnelson0 [already has a PR to fix it](https://github.com/fiddyspence/puppetlint-variablecase/pull/2) so we switched to his fork until the author has a chance to merge the PR.

```
[16:05][mmckinst] bundle exec gem dependency --reverse-dependencies puppet-lint | fgrep '(puppet-lint '
    puppet-lint-empty_string-check-0.2.2 (puppet-lint (< 3.0, >= 1.0))
    puppet-lint-leading_zero-check-0.1.1 (puppet-lint (< 3.0, >= 1.0))
    puppet-lint-unquoted_string-check-0.3.0 (puppet-lint (< 3.0, >= 1.0))
    puppet-lint-variable_contains_upcase-1.0.0 (puppet-lint (~> 1.0))
    puppet-lint-version_comparison-check-0.2.1 (puppet-lint (< 3.0, >= 1.0))
    puppetlabs_spec_helper-1.1.1 (puppet-lint (>= 0))
[16:05][mmckinst] 
```

To prevent a regression of this, should you add another gem that depends on puppet-lint ~>1.0, add the following to your Gemfile:

```
gem 'puppet-lint', '~>2.0'
```

**Comment by mmckinst** [comment]
Confirmation it works as desired:

```
$ cat cat -b test.pp 
     1	class test {
     2	  $good = $facts['puppetversion']
     3	  $bad = $puppetversion
     4	  $standard = $::puppetversion
     5	}
$ bundle exec puppet-lint --version && bundle exec puppet-lint test.pp
puppet-lint 2.0.0
ERROR: test not in autoload module layout on line 1
WARNING: top-scope variable being used without an explicit namespace on line 3
WARNING: class not documented on line 1
$
```

**Comment by rnelson0** [comment]
@mmckinst Thanks for the archeology and confirmation! I will close this out but everyone can feel free to re-open or open a new ticket if there are outstanding issues from this. Thanks!

**Comment by tehmaspc** [comment]
@mmckinst rock! thanks bud; 

cheers all!
